### PR TITLE
Refactor config.pl to make it easier to add presets

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
This PR changes `scripts/config.pl`:

* New preset `crypto_all`, which is what I set out to do.
* Refactor the code to make it easier to add presets.
* New script `tests/scripts/test-config-script.pl` which can be used to see how changes to `config.pl` change its output.
* Enable `MBEDTLS_PLATFORM_NV_SEED_ALT` in the full configuration.

This should be backported to make the config scripts uniform, I'll do that once the reviews are in.

### CODE REVIEW

File: scripts/config.pl [hunk 1]
```diff
@@ -131,12 +173,19 @@ sub disjunction_re {
...
 my $no_exclude_baremetal_re = disjunction_re(@non_excluded_baremetal);
+my $debug_features_re = disjunction_re(@debug_features);
+my $x509_features_re = disjunction_re(@x509_features);
+my $tls_features_re = disjunction_re(@tls_features);
+
 my %presets = (
                baremetal => {exclude_re => qr/($exclude_re|$exclude_baremetal_re)/,
                              no_exclude_re => qr/(?!$exclude_baremetal_re)$no_exclude_re|$no_exclude_baremetal_re/},
+               crypto_all => {exclude_re => qr/$exclude_non_full_re|$x509_features_re|$tls_features_re|$debug_features_re/,
+                              no_exclude_re => qr/^$/},
```
Comment (Patater) [inline scripts/config.pl]: 1. Are you happy with the resulting config.h? I see we still have `MBEDTLS_ASN1_PARSE_C`, `MBEDTLS_ASN1_WRITE_C`, `MBEDTLS_PEM_PARSE_C`, `MBEDTLS_PK_PARSE_C`, `MBEDTLS_PK_C`, and `MBEDTLS_BASE64_C` enabled, for instance. Are those all really wanted?

2. Should we include the resulting config.h in our set of pre-made example config files (in `configs`)?

File: scripts/config.pl [hunk 2]
```diff
@@ -114,6 +118,44 @@
...
 );
 
+# Things that should be excuded in non-full configurations
+my @excluded_non_full = qw(
+MBEDTLS_AES_ROM_TABLES
+MBEDTLS_CAMELLIA_SMALL_MEMORY
+MBEDTLS_DEPRECATED_\w+
+MBEDTLS_ENTROPY_FORCE_SHA256
+MBEDTLS_ENTROPY_NV_SEED
+MBEDTLS_HAVEGE_C
```
Comment (Patater) [inline scripts/config.pl]: Why not include `HAVEGE_C` in `crypto_all`? Don't have the time? :p

File: scripts/config.pl [hunk 3]
```diff
@@ -114,6 +118,44 @@
 MBEDTLS_PLATFORM_FPRINTF_ALT
 );
 
+# Things that should be excuded in non-full configurations
+my @excluded_non_full = qw(
+MBEDTLS_AES_ROM_TABLES
```
Comment (Patater) [inline scripts/config.pl]: Why not include AES ROM tables in `crypto_all`?

File: scripts/config.pl [hunk 4]
```diff
@@ -114,6 +118,44 @@
...
+MBEDTLS_CAMELLIA_SMALL_MEMORY
+MBEDTLS_DEPRECATED_\w+
+MBEDTLS_ENTROPY_FORCE_SHA256
+MBEDTLS_ENTROPY_NV_SEED
+MBEDTLS_HAVEGE_C
+MBEDTLS_MEMORY_\w+
+MBEDTLS_PADLOCK_C
```
Comment (Patater) [inline scripts/config.pl]: We exclude Padlock. Should we also exclude `MBEDTLS_AESNI_C`?

File: scripts/config.pl [hunk 5]
```diff
@@ -115,8 +115,11 @@
 );
 
 # Things that should be enabled in "full" even if they match @excluded
+# These are all the platform ALT definitions, so that we test them in "full",
+# except MBEDTLS_PLATFORM_SETUP_TEARDOWN_ALT which requires a platform_alt.h
+# that we do not provide.
```
Comment (Patater) [inline scripts/config.pl]: Should we provide one for testing or would that not be valuable? I'm trying to judge if we should file an issue for plugging this test hole.

**Comment by sbutcher-arm** [comment]
@gilles-peskine-arm - What are your plans for this PR? It's original purpose was to support the Crypto refactoring work for PSA, but now that's moving into the submodule, I'm not sure this PR makes as much sense.

Do you plan to rework it following the reviews? And to fix the merge conflicts?

**Comment by gilles-peskine-arm** [comment]
Now that crypto and TLS are becoming separate products, I no longer have a use for this PR. I still think it would be useful but it's low priority and I don't intend do a rebase. The work is there if someone wants to pick it up, but I'm closing this PR.

**Pull request closed without merging**